### ISSUE TITLE
radial_average

### ISSUE THREAD

**Issue by geoffwoollard** [mjo22/cryojax#133]
I wanted to whiten a particle stack and was hoping to use the `WhiteningFilter`.

I did some sanity checks on random gaussian noise (already white), and suspect there is a bug in `radial_average`.

When I radially average the spectra, I don't get back a flat filter. Of course there is some sampling noise here since the image (here 64 x 64) is finite. But even in expectation (1000 samples), it's not flat.

```python
shape = (64,64)
k_min, k_max = 0, 0.5
pixel_size=1
k_step = 1.0 / (pixel_size * max(*power.shape))
bins = jnp.arange(k_min, k_max, k_step)
frequency_grid = np.fft.fftshift(make_frequencies(shape, grid_spacing=1, half_space=False))
radial_frequency_grid = jnp.linalg.norm(frequency_grid, axis=-1)

freqs = frequency_grid
interpolating_radial_frequency_grid = jnp.linalg.norm(freqs, axis=-1)

n_samples=10000
spectra = np.empty((n_samples,*shape))
for idx, seed in enumerate(range(n_samples)):
    np.random.seed(seed)
    image_r = np.random.normal(np.zeros(shape), 1)
    image_f = np.fft.fftshift(np.fft.fftn(image_r))

    power = (image_f * jnp.conjugate(image_f)).real
    spectrum = radial_average(
        power, radial_frequency_grid, bins, interpolating_radial_frequency_grid
    )
    spectra[idx] = spectrum
```

```
plt.imshow(spectra.mean(0))
plt.colorbar()
plt.title(f'Expectation of radially averaged power of random gaussian noise \n n_samples={n_samples}')
```
![0b82604a-9683-47e6-bbae-9117ec11f6bf](https://github.com/mjo22/cryojax/assets/4852692/9fc475ac-471a-46f9-a381-e660a028936e)

```
plt.imshow(spectra.std(0))
plt.colorbar()
plt.title(f'Std of radially averaged power of random gaussian noise \n n_samples={n_samples}')
```
![c13c2f64-20e5-45ec-9861-78c138112918](https://github.com/mjo22/cryojax/assets/4852692/87d2441f-a8db-4b89-a407-fbb26289e6f7)

Using code from commit b39bd0de800692503d806844c45fee76ceaf18a0

**Comment by geoffwoollard** [comment]
Here is just radially averaged gaussian noise

This would reflect the expectation of the sample average (averaging over more pixels in a larger radius --> higher probability of being close to zero). There's no ring artefacts, just a gradual decay, which makes sense.

```
n_samples=10000
shape = (64, 64)
radial_averages = np.empty((n_samples,*shape))

coordinates = make_coordinates(shape)
radial_grid = jnp.hypot(coordinates[:,:,0], coordinates[:,:,1])
k_min = 0
k_max = radial_grid.max()
pixel_size=1
k_step = 1 #1.0 / (pixel_size * max(*shape))
bins = jnp.arange(k_min, k_max, k_step)
interpolating_radial_grid = radial_grid

for idx, seed in enumerate(range(n_samples)):
    np.random.seed(seed)
    image_r = np.random.normal(np.zeros(shape), 1)

    radial_average_image = radial_average(
        image_r, radial_grid, bins, interpolating_radial_frequency_grid
    )
    radial_averages[idx] = radial_average_image
```

```
plt.imshow(radial_averages.mean(0))
```
![a058be0f-9d50-4749-b215-1bee0692c234](https://github.com/mjo22/cryojax/assets/4852692/28f2932c-759f-4275-8242-0834ca7a0503)

```plt.imshow(radial_averages.std(0))
```
![be6004db-7bc8-4c53-a476-f96b94123b45](https://github.com/mjo22/cryojax/assets/4852692/3e5d6bc5-393f-4036-a3be-c32db0fba111)

**Comment by mjo22** [comment]
This is a really good check to have, thanks for testing this. 

I actually just added some big improvements and fixed some bugs in the power spectrum. See the most recent commits to main—maybe this will address the issue. The “interpolated_radial_frequency_grid” argument is deprecated because I naively thought I could downsample a power spectrum profile to an arbitrary grid. Instead, I now more simply just put it on the same grid it was computed on. Downsampling can then be done by cropping in real space. It’s also much faster now.

However, I’m not sure this will fix the issue. Can you

1) Show what these plots look like with the new code, along with a plot of the radial profile?
2) Normalize the power spectrum by the number of modes? Note the image scale bars are very close to 64*64 = 4096. I believe the radial profile you plotted for the power is basically just small fluctuations around 1.

So I think that there may be no bug, despite the expectation. I think that it will converge to a flat spectrum as 1 / sqrt(n_samples), so there may not be enough samples. Regardless, the imshow will be misleading, especially without normalization.

**Comment by geoffwoollard** [comment]
I'm getting different and expected behaviour for the new radial_average function. It also more-or-less agrees with a numpy function that colab generative ai autocompleted.

https://colab.research.google.com/gist/geoffwoollard/e4b6bf14f4b1813d0b2d4a201c766923/radial_average.ipynb

**Comment by mjo22** [comment]
It looks like the AI implementation is the equivalent of `right = False` in the digitize call, while I ended up finding bugs in the average if I used this option. Do they look more similar if you change the AI option to the equivalent of `right = True`?

**Comment by mjo22** [comment]
Closing this as resolved. Need to still add tests for the whitening filter though
